### ISSUE TITLE
Nested External Schema Reference doesn't work

### ISSUE THREAD

**Issue by OllieJennings** [mafintosh/is-my-json-valid#67]
When using this library, the ref to an external schema just doesn't work, but when using others such as tv4, it works perfectly fine.

The Schema
```
{
  "type": "object",
  "properties": {
     "body":  "$ref": "#oauth"
  }
}
```

And the defs schema
```
{
  "definitions": {
    "oauth": { ... }
  }
}
```

when doing

``
validate = validator(schema, { "schemas": defs.definitions });
validate(req);
``

The error l get is: `reference schema does not match`

**Comment by OllieJennings** [comment]
From playing around with the code, it seems that you do find the referenced schema within the if statement `if(node.$ref)` and that the `sub` value within that is equal to the referenced schema. However you still return an error.

TLDR: code locates referenced schema, but doesn't include it, therefore thinks it's missing

**Comment by OllieJennings** [comment]
So it turns out, after a lot of investigating, that the example of external reference schema on the main page works, however if you were to nest it, such as:

```
var schema = {
  "body": {
    "$ref": '#ext'
  }
};
```

and then:

```
var extSch = {
  "definitions": {
    "ext": {
      "type": "object",
      "required": ["log"],
      "properties": {
        "log": { "type": "string" }
      }
    }
  }
}
```

Then you get the following results:

```
validate({ "body": { "log": "hello" } }); // returns true
validate('hello'); // returns true
validate(42); // returns true
```

Which is a bug, as it can't seem to pull in and actually use the referenced schema

**Comment by OllieJennings** [comment]
@mafintosh is this a bug, or is this standard json schema functionality?

**Comment by mafintosh** [comment]
could check this against the latest version? @chrahunt fixed some stuff with external schema handling

**Comment by chrahunt** [comment]
@OllieJennings This looks like standard JSON Schema functionality. JSON Schema doesn't implicitly require properties that are present in the schema so as it is, your schema doesn't place any restrictions on the format of objects that it will match. If you wanted to match objects of the form `{ "body": { "log": "hello" } }` you would use the schema:

```json
{
  "type": "object",
  "properties": {
    "body": { "$ref": "#ext" }
  },
  "required": ["body"]
}
```

The following script behaves as-expected:

```javascript
var validator = require('is-my-json-valid');

var schema = {
  "type": "object",
  "properties": {
    "body": { "$ref": "#ext" }
  },
  "required": ["body"]
};

var extSch = {
  "definitions": {
    "ext": {
      "type": "object",
      "required": ["log"],
      "properties": {
        "log": { "type": "string" }
      }
    }
  }
};

var validate = validator(schema, {
  schemas: {
    ext: extSch.definitions
  }
});

console.log(validate({ "body": { "log": "hello" } })); // returns true
console.log(validate('hello')); // returns false
console.log(validate(42)); // returns false
```

**Comment by OllieJennings** [comment]
ahh @chrahunt thank you so much for this, cleared up a lot of things for me!

**Comment by OllieJennings** [comment]
@chrahunt The only slight bug in the above l have found is that if you do:

```
validate({ "body": { "log": [] }}); // returns true but should be false
validate({ "body": { "log": 20 }}); // returns true as well but should be false 
```

In the above, the `log` property should be a string, so passing an empty array or an integer should cause an error.

**Comment by chrahunt** [comment]
My fault, I made a typo in my previous comment. I've adjusted the code so that it references the `ext` properly. Just a note, for small schemas, it's good practice to keep all of the items together in the same schema unless they're needed in multiple schemas. Going from your example:

```javascript
var schema = {
  "type": "object",
  "definitions": {
    "ext": {
      "type": "object",
      "required": ["log"],
      "properties": {
        "log": { "type": "string" }
      }
    }
  },
  "properties": {
    "body": { "$ref": "#/definitions/ext" }
  },
  "required": ["body"]
};
```

works, and "#/definitions/ext" is found automatically by the library, no need to use the `schemas` option. If you're working with multiple schemas and have duplicated pieces, then moving those to a file is one way to go. You can see an example of this [here](https://github.com/chrahunt/tpr-schema-test/tree/master/schemas/2), where `team` and `time` are in `definitions.json` so they can be referenced by `player.json`, `info.json`, and `data.json`.  The player and data files have examples of the same-schema `$ref` usage I mentioned above. Looking in those three files you can also see how the reference is made to the items contained in `definitions.json`. Those three files themselves are referenced in `replay.json`. Then (using v2.12.0 of `is-my-json-valid`), we can construct a validate function for the `replay` schema just like this:

```javascript
var validator = require("is-my-json-valid");
var validate = validator(require("./replay.json"), {
  schemas: {
    "player.json": require("./player.json"),
    "info.json": require("./info.json"),
    "data.json": require("./data.json"),
    "definitions.json": require("./definitions.json")
  }
});
```

For more information on JSON Schema, [this](http://spacetelescope.github.io/understanding-json-schema/index.html) is a good resource, and they have information on structuring complex schemas [here](http://spacetelescope.github.io/understanding-json-schema/structuring.html).
